packages/grid/x-data-grid/src/hooks/utils/useTimeout.ts

```diff
@@ -0,0 +1,34 @@
+import { useLazyRef } from './useLazyRef';
+import { useOnMount } from './useOnMount';
+
+class Timeout {
+  static create() {
+    return new Timeout();
+  }
+
+  currentId: number = 0;
+
+  start(delay: number, fn: Function) {
+    this.clear();
+    this.currentId = setTimeout(fn, delay);
+  }
+
+  clear = () => {
+    if (this.currentId !== 0) {
+      clearTimeout(this.currentId);
+      this.currentId = 0;
+    }
+  };
+
+  disposeEffect = () => {
+    return this.clear;
+  };
+}
+
+export function useTimeout() {
+  const timeout = useLazyRef(Timeout.create).current;
+
+  useOnMount(timeout.disposeEffect);
```

Comment:
`useOnMount` has a clear semantic. It also doesn't re-allocate the `[]` object on every render, so less work for the GC. Very small cost again, but these add up and end up creating more frequent and longer GC pauses if we allocate objects for no reason.

Reply:
I'm a bit worried that 1. It could be hard for team members to learn to use it. 2. When reviewing the code, it's someone new reviewers need to learn 3. I wouldn't be surprised if it will be missed, and even once known, the old habits will persist.

But I won't push against it, why not give it a try.

Reply:
We had that discussion when I added `useOnMount` in some older PR, but since I've seen other members of the team [adopt it](https://github.com/mui/mui-x/pull/9712#discussion_r1266935336), so I think it's beneficial. React hooks are a bit raw, having a small set of primitives added on top of them isn't a bad idea, imho.